cryptography/hedera-cryptography-rpm/src/main/rust/common/src/address_book.rs

```
@@ -0,0 +1,110 @@
+use derive_more::derive::Deref;
+use serde::{Deserialize, Serialize};
+use serde_big_array::Array;
+use smallvec::SmallVec;
+
+use crate::ed25519;
+
+pub type Weight = u64;
+
+pub const MAXIMUM_VALIDATORS: usize = 64;
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct AddressBookEntry {
+    pub ed25519_public_key: Array<u8, { ed25519::PUBLIC_KEY_LENGTH }>,
+    pub weight: Weight,
+}
+
+#[repr(transparent)]
+#[derive(Debug, Deref, Clone, Serialize, Deserialize)]
+pub struct AddressBook(pub SmallVec<[AddressBookEntry; MAXIMUM_VALIDATORS]>);
+
+impl AddressBook {
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn get(&self, index: usize) -> Option<&AddressBookEntry> {
+        self.0.get(index)
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &AddressBookEntry> {
+        self.0.iter()
+    }
+
+    pub fn new(verifying_keys: Vec<ed25519::VerifyingKey>, weights: Vec<u64>) -> Self {
+        assert!(
+            verifying_keys.len() <= MAXIMUM_VALIDATORS,
+            "Too many verifying keys"
+        );
+        assert!(
+            verifying_keys.len() == weights.len(),
+            "Different number of verifying keys and weights"
+        );
```

Minor: I presume this will panic, so we'd have to validate the inputs before calling this constructor, so that we can return an error to the Java layer. For this reason, it would be nice to have this validation implemented in a dedicated static method returning a boolean. This constructor would call the method and panic if it returns false. The JNI wrapper would call the method and simply return an error return value to Java immediately, thus preventing the panic.

---

Do you mean that the rust function can assume valid inputs, where lengths match for example?

---

I suppose it shouldn't just assume it and should panic still. I just suggest to extract the validation into a separate method that simply returns a boolean instead of causing a panic. This would allow the quoted call to call it and then panic explicitly, and it would also allow the JNI wrapper to also call the validation method and simply return an error code back to Java w/o a panic.

---

Makes sense. However, for this case, since the check is quite straightforward, is it better to simply have JNI bridge enforce it and return error to Java if the condition is not met? This may keep the rust-side interface small.